File: spec/cpp_stl_98/test_str_literals.cpp
#include <boost/test/unit_test.hpp>

#include <str_literals.h>

#include <iostream>
#include <fstream>
#include <vector>

BOOST_AUTO_TEST_CASE(test_str_literals) {
    std::ifstream ifs("src/fixed_struct.bin", std::ifstream::binary);
    kaitai::kstream ks(&ifs);
    str_literals_t* r = new str_literals_t(&ks);

    BOOST_CHECK_EQUAL(r->complex_str(), std::string("\x00\x01\x02\x07\x08\x0a\x0d\x09\x0b\x0c\x1b\x3d\x07\x0a\x24\u263b", 18));
    BOOST_CHECK_EQUAL(r->double_quotes(), "\x22\x22\x22");
    BOOST_CHECK_EQUAL(r->backslashes(), "\x5c\x5c\x5c");
    BOOST_CHECK_EQUAL(r->octal_eatup(), std::string("\x00\x32\x32", 3));
    BOOST_CHECK_EQUAL(r->octal_eatup2(), std::string("\x02\x32", 2));

    delete r;
}

File: spec/construct/test_str_literals.py
import unittest

from str_literals import _schema

class TestStrLiterals(unittest.TestCase):
    def test_str_literals(self):
        r = _schema.parse_file('src/fixed_struct.bin')

        self.assertEqual(self.str_to_arr(r.complex_str), [0, 1, 2, 7, 8, 10, 13, 9, 11, 12, 27, 61, 7, 10, 36, 9787])
        self.assertEqual(self.str_to_arr(r.double_quotes), [34, 34, 34])
        self.assertEqual(self.str_to_arr(r.backslashes), [92, 92, 92])
        self.assertEqual(self.str_to_arr(r.octal_eatup), [0, 50, 50])
        self.assertEqual(self.str_to_arr(r.octal_eatup2), [2, 50])

    def str_to_arr(self, s):
        return [ord(c) for c in s]
